pkg/controllers/management/auth/globalroles/globalrole_handler.go

```diff
@@ -196,6 +217,110 @@ func (gr *globalRoleLifecycle) reconcileCatalogRole(globalRole *v3.GlobalRole) e
 	return nil
 }
 
+// reconcileNamespacedRoles ensures that Roles exist in each namespace of NamespacedRules
+func (gr *globalRoleLifecycle) reconcileNamespacedRoles(globalRole *v3.GlobalRole) error {
+	var returnError error
+
+	// For collecting all the roles that should exist for the GlobalRole
+	roleUIDs := []types.UID{}
+
+	for ns, rules := range globalRole.NamespacedRules {
+		_, err := gr.nsLister.Get("", ns)
+		if err != nil {
+			returnError = multierror.Append(returnError, errors.Wrapf(err, "couldn't get namespace %s", ns))
+			continue
+		}
+		// Check if the role exists
+		roleName := globalRole.Name + "-" + ns
+		role, err := gr.rLister.Get(ns, roleName)
+		if err != nil {
+			if !apierrors.IsNotFound(err) {
+				returnError = multierror.Append(returnError, err)
+				continue
+			}
+			newRole := &v1.Role{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      roleName,
+					Namespace: ns,
+					OwnerReferences: []metav1.OwnerReference{
+						{
+							APIVersion: globalRole.APIVersion,
+							Kind:       globalRole.Kind,
+							Name:       globalRole.Name,
+							UID:        globalRole.UID,
+						},
+					},
+					Labels: map[string]string{
+						grOwnerLabel: wrangler.SafeConcatName(globalRole.Name),
+					},
+				},
+				Rules: rules,
+			}
+			createdRole, err := gr.rClient.Create(newRole)
+			if err == nil {
+				roleUIDs = append(roleUIDs, createdRole.UID)
+				continue
+			} else if !apierrors.IsAlreadyExists(err) {
+				returnError = multierror.Append(returnError, err)
+				continue
+			} else {
+				// In the case that the role already exists, we fall through to check that the rules are correct
+				role = newRole
+			}
+		}
+		if role != nil {
+			roleUIDs = append(roleUIDs, role.GetUID())
+
+			// Check that the rules for the existing role are correct and that it has the right Owner Label
+			if !reflect.DeepEqual(role.Rules, rules) || role.Labels[grOwnerLabel] != globalRole.Name {
+				newRole := role.DeepCopy()
+				newRole.Rules = rules
+				if newRole.Labels == nil {
+					newRole.Labels = map[string]string{}
+				}
+				newRole.Labels[grOwnerLabel] = globalRole.Name
+
+				_, err := gr.rClient.Update(newRole)
+				if err != nil {
+					returnError = multierror.Append(returnError, err)
+					continue
+				}
+			}
+		}
+	}
+
+	// get all the roles claiming to be owned by this GR and remove any that shouldn't exist
+	r, err := labels.NewRequirement(grOwnerLabel, selection.Equals, []string{globalRole.Name})
+	if err != nil {
+		return multierror.Append(returnError, errors.Wrapf(err, "couldn't create label: %s", grOwnerLabel))
+	}
+
+	roles, err := gr.rLister.List("", labels.NewSelector().Add(*r))
```

Comment:
This strikes me as a bit inefficient - in the standard case (no corrupt roles) we are going to individually get each item and then re-list them again here. Did you consider any other implementations (like processing the return of a single list call)?

Reply:
So giving it some thought, I see 2 options to improve efficiency:

1. I create a map `m` of roles with the matching grOwner label. As I go through the namespaces in NamespacedRules, if I find a role that's in `r`, remove it. After the function, I do `purgeInvalidNamespacedRoles` on the remaining items
2. Using the current method (but with a map of uids instead of a slice), before I call `purgeInvalidNamespacedRoles` I check the length of `roles` and `uids`. If they're the same size, I don't need to call `purgeInvalidNamespacedRoles` because they are guaranteed to be the same list.

In case 1, I maintain the map of roles with the matching labels throughout. In case 2 I only look through it if there are invalid Roles that need to be deleted. So in the case that this is called most frequently on GRs with no invalid roles, case 2 would require fewer operations. In the case where there are invalid roles, case 1 is more efficient.

Reply:
I personally think that since the majority of the time this is called, there will be no invalid roles, case 2 is preferable. However the savings either way are minute. We're talking about minimizing map lookups and deletes which are O(1) time so I don't feel too strongly either way.